ls1intum/Artemis
# `Plagiarism detection`: Add limited cpc attempts

### Checklist
⚠️ **This PR targets `cpc` branch instead of `develop`.** The reason is that there is an issue with `@Scheduled` annotations on test servers which prevents cpc related changes to be tested. **All changes in this PR are suitable to be easily tested locally.** Once this PR gets approvals from the team and the maintainer (@MarkusPaulsen) it will be merged into `cpc`. Once issues with test servers are fixed @jakubriegel and @MarkusPaulsen will test everything e2e. @jakubriegel keeps `cpc` up to date with `develop`.

#### General
#### Server
#### Client
#### Changes affecting Programming Exercises

Explicit tests not necessary as this feature operates on generic exercise abstraction.

### Motivation and Context
This PR is part of _continuous plagiarism control_ project. It implements CPC-9 milestone from @jakubriegel's thesis. More about that can be discovered in the thesis proposal: https://confluence.ase.in.tum.de/display/ArTEMiS/MA+Jakub+Riegel?preview=/157431627/166332701/riegel_proposal.pdf

The cpc runs plagiarism checks automatically each night. Students may try to fool the checks but sending additional submissions and learning how the plagiarism checks mechanism behaves. To prevent this instructors should have a possibility to limit the maximal number of submissions to be tested.

### Description
This PR introduces the possibility to limit the maximal number of submissions to be tested by cpc. Any submission that exceed the limit will be accepted but cpc will reduce it score to 0. The feature allows students to submit further solutions and dispute the plagiarism suspicion after the due date.

Changes:
- server changes adding new logic to `ContinuousPlagiarismControlResultsService`
- client changes adding new input to the exercise create/edit form of text, modeling and programming exercises

### Steps for Testing
Prerequisites:
- ⚠️ **Please test this change locall**y (no ci setup is needed, see the note at the top for more details)
- 1 Instructor
- 2 Students
- 1 modeling or text exercise

1. Change cron so that cpc is triggered every 30s:
    - find the line `@Scheduled(cron = "${artemis.scheduling.continuous-plagiarism-control-trigger-time:0 0 4 * * *}")`
    - replace it with `@Scheduled(initialDelay = 30_000, fixedDelay = 30_000)`
1. Create an exercise with
    - due date in the future
    - selected: _Enable Continuous Plagiarism Control_
    - `2` or your own number in: _Max. number of Continuous Plagiarism Control detections_
1. Create two identical submissions from two different users **and wait 1min**
1. Both submissions should have received results with feedback and 0 score
1. Wait 1 more minute
1. Both submissions should have received additional results with same comment
1. Change one submission by 1 letter (so that it is plagiarism but it is updated) **and wait 1min**
1. Both submissions receive the same feedback
1. Change the same submission by 1 letter again **and wait 1min**
1. If you selected `2` as _Max. number of Continuous Plagiarism Control detections_ then changed submission should get feedback with information that no more submissions are possible.
1. Change the same submission so that it is completely different **and wait 1min**
1. Changed submission should get feedback with information that no more submissions are possible.

### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests


### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ContinuousPlagiarismControlResultsService.java | 100% | ✅                           |


### Screenshots
New input in exercise create/update form with default value of 3:
<img width="1504" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/17221fd1-5546-43c2-94c2-f274b57fb340">
